hazelcast/src/main/java/com/hazelcast/sql/impl/expression/predicate/ComparisonPredicate.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.expression.predicate;
+
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.sql.impl.SqlDataSerializerHook;
+import com.hazelcast.sql.impl.expression.BiExpression;
+import com.hazelcast.sql.impl.expression.Expression;
+import com.hazelcast.sql.impl.expression.ExpressionEvalContext;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.type.QueryDataType;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Implements evaluation of SQL comparison predicates.
+ *
+ * @see ComparisonMode
+ */
+public final class ComparisonPredicate extends BiExpression<Boolean> implements IdentifiedDataSerializable {
```

Comment:
It seems that the comparison of objects is not supported. When I attempt to do `f=f` for `t(f:ANY)`, the following exception is thrown:
```
Cannot apply '=' to arguments of type '<ANY> = <ANY>'. Supported form(s): '<COMPARABLE_TYPE> = <COMPARABLE_TYPE>'
```
There are two potential problems here:
1. Why object comparison is not supported in the first place?
1. If there is a reason to not support it right now, then error message is not intuitive: `ANY` could be `COMPARABLE_TYPE`

Reply:
created an issue https://github.com/hazelcast/hazelcast/issues/17301 for 4.2

Reply:
created an issue https://github.com/hazelcast/hazelcast/issues/17302 for 4.2